Sources/Apodini/Modifier/ResponseTransformer.swift

```diff
@@ -0,0 +1,88 @@
+    /// Both types (`ResponseTransformable` and `TransformedContent`) have to conform to `Encodable`.
+    /// - Parameter response: The response that should be transformed
+    func transform(content: Self.InputContent) -> Self.Content
+}
+
+internal struct ResponseTransformingHandler<D, T>: Handler where D: Handler, T: ResponseTransformer, D.Response.Content == T.InputContent {
+    let transformed: Delegate<D>
+    let transformer: Delegate<T>
+    
+    @Environment(\.connection) var connection
+    
+    func handle() throws -> EventLoopFuture<Response<T.Content>> {
+        try transformed().handle().transformToResponse(on: connection.eventLoop).flatMapThrowing { responseToTransform in
+            try responseToTransform.map { content in
+                try transformer().transform(content: content)
+            }
+        }
+    }
+}
+
+public struct ResponseTransformingHandlerInitializer<T: ResponseTransformer>: DelegatingHandlerInitializer {
+    public typealias Response = Apodini.Response<T.Content>
+    
+    let transformer: T
+    
+    public func instance<D>(for delegate: D) throws -> SomeHandler<Response> where D: Handler {
+        if let transformingHandler =
+            (TransformerCandidate(
+                transformer: transformer,
+                delegate: delegate) as? Transformable)?() as? SomeHandler<Response> {
+            return transformingHandler
+        }
+        
+        fatalError("Cannot use response transformer \(transformer) with handler of type \(D.self) because content types do not match.")
+    }
+}
+
+private struct TransformerCandidate<Transformer: ResponseTransformer, Delegate: Handler> {
+    let transformer: Transformer
+    let delegate: Delegate
+}
+
+extension TransformerCandidate: Transformable where Transformer.InputContent == Delegate.Response.Content {
+    func callAsFunction() -> Any {
+        SomeHandler<Response<Transformer.Content>>(ResponseTransformingHandler<Delegate, Transformer>(
+                                                    transformed: Apodini.Delegate(delegate),
+                                                    transformer: Apodini.Delegate(transformer)))
+    }
+}
+
+private protocol Transformable {
+    func callAsFunction() -> Any
+}
+
+extension Handler {
+    /// A `response` modifier can be used to transform the output of a `Handler`'s response to a different type using a `ResponseTransformer`
+    /// - Parameter responseTransformer: The `ResponseTransformer` used to transform the response of a `Handler`
+    /// - Returns: The modified `Handler` with a new `ResponseTransformable` type
+    public func response<T: ResponseTransformer>(
+        _ responseTransformer: T
+    ) -> DelegationModifier<Self, ResponseTransformingHandlerInitializer<T>> where Self.Response.Content == T.InputContent {
+        self.delegated(by: ResponseTransformingHandlerInitializer(transformer: responseTransformer))
```

Comment:
Regarding your type question. The problem really is that the input parameter of type `D` of the `instance(for:)` is completely decoupled from the initialization going on here.
I don't think you can "solve" this completely, only really option would be trying to make it "less" "ugly"?
Can you somehow pass the generic type `Self` over to the `ResponseTransformingHandlerInitializer` (with a `where` enforcing that output type equals inout type)? Then you could just check if the both generic types are still the same in `instance(for:)`, but not sure if this makes anything better 🤷

Reply:
Yes, exactly - it is just about about making it less ugly. But I don't thing what you suggest would lead to a solution. It isn't nice, but it works, so I'll leave it as is for now.